File: Windows/app/Native/MeterCoreNative.cs
using System.Runtime.InteropServices;

namespace TollCat;

/// <summary>
/// 和 Android <c>MeterCoreNative.kt</c> 同构：一个导出函数一个方法，
/// JSON 字符串进出，反序列化只在这一层之上。C# 侧零网络请求。
/// </summary>
internal static partial class MeterCoreNative
{
    private const string Lib = "MeterCoreCLR";

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial void tollcat_set_resource_root(string path);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_catalog(string localeTag);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_setup_guide(string providerId, string localeTag);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_format_usd(string raw, string currency, string localeTag);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_format_month_and_day(long millis, string localeTag);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_dashboard(
        string snapshotsJson,
        string subscriptionsJson,
        long nowMillis,
        string currency,
        string localeTag,
        string filterJson);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_history_chart(
        string providerId,
        string snapshotsJson,
        string range,
        long nowMillis);

    [LibraryImport(Lib)]
    private static partial nint tollcat_design_seed(long nowMillis);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_fetch(string providerId, string fieldsJson, long nowMillis);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_post_usage(string payloadJson);

    [LibraryImport(Lib, StringMarshalling = StringMarshalling.Utf8)]
    private static partial nint tollcat_post_feedback(string payloadJson);

    [LibraryImport(Lib)]
    private static partial void tollcat_free(nint ptr);

    public static void SetResourceRoot(string path) => tollcat_set_resource_root(path);

    public static string CatalogJson(string localeTag) => Take(tollcat_catalog(localeTag));

    public static string SetupGuideJson(string providerId, string localeTag) =>
        Take(tollcat_setup_guide(providerId, localeTag));

    public static string FormatUsd(string raw, string currency, string localeTag) =>
        Take(tollcat_format_usd(raw, currency, localeTag));

    public static string FormatMonthAndDay(long millis, string localeTag) =>
        Take(tollcat_format_month_and_day(millis, localeTag));

    public static string ComputeDashboardJson(
        string snapshotsJson,
        string subscriptionsJson,
        long nowMillis,
        string currency,
        string localeTag,
        string filterJson) =>
        Take(tollcat_dashboard(snapshotsJson, subscriptionsJson, nowMillis, currency, localeTag, filterJson));

    public static string HistoryChartJson(string providerId, string snapshotsJson, string range, long nowMillis) =>
        Take(tollcat_history_chart(providerId, snapshotsJson, range, nowMillis));

    public static string DesignSeedJson(long nowMillis) => Take(tollcat_design_seed(nowMillis));

    public static string FetchProviderJson(string providerId, string fieldsJson, long nowMillis) =>
        Take(tollcat_fetch(providerId, fieldsJson, nowMillis));

    public static string PostUsageJson(string payloadJson) => Take(tollcat_post_usage(payloadJson));

    public static string PostFeedbackJson(string payloadJson) => Take(tollcat_post_feedback(payloadJson));

    private static string Take(nint ptr)
    {
        if (ptr == nint.Zero) return "{}";
        try
        {
            return Marshal.PtrToStringUTF8(ptr) ?? "{}";
        }
        finally
        {
            tollcat_free(ptr);
        }
    }

}

File: Android/app/src/main/kotlin/com/zhechengqi/tollcat/MeterCoreNative.kt
package com.zhechengqi.tollcat

import android.content.Context
import java.io.File

/**
 * JNI 入口。库加载顺序按 DT_NEEDED 来：ART 不会按 LD_LIBRARY_PATH 解析 Swift runtime。
 */
object MeterCoreNative {
    @Volatile
    var loaded = false
        private set

    fun load(context: Context) {
        if (loaded) return
        synchronized(this) {
            if (loaded) return
            val dir = File(context.applicationInfo.nativeLibraryDir)
            val ordered = arrayOf(
                "libz.so",
                "libc++_shared.so",
                "libswiftCore.so",
                "libswiftSwiftOnoneSupport.so",
                "libswiftSynchronization.so",
                "libswift_Concurrency.so",
                "libswift_RegexParser.so",
                "libswift_StringProcessing.so",
                "libswiftRegexBuilder.so",
                "libswift_Builtin_float.so",
                "libswift_math.so",
                "libswift_Volatile.so",
                "libswift_Differentiation.so",
                "libBlocksRuntime.so",
                "libdispatch.so",
                "libswiftDispatch.so",
                "libswiftAndroid.so",
                "libswiftDistributed.so",
                "libswiftObservation.so",
                "libFoundationEssentials.so",
                "lib_FoundationICU.so",
                "libFoundationInternationalization.so",
                "libFoundation.so",
                "libFoundationNetworking.so",
                JNI_LIB,
            )
            for (name in ordered) {
                val file = File(dir, name)
                if (file.exists()) {
                    System.load(file.absolutePath)
                }
            }
            loaded = true
        }
    }

    @JvmStatic
    external fun setResourceRoot(path: String)

    @JvmStatic
    external fun formattedDesignSeed(): String

    @JvmStatic
    external fun designSeedMatches(): Boolean

    @JvmStatic
    external fun fixtureProofJson(): String

    @JvmStatic
    external fun stubFetchCloudflareJson(): String

    @JvmStatic
    external fun catalogJson(localeTag: String): String

    /** 向导整篇教程 + 行内深链解析表。目录里没有这家时返回 {"missing":true}。 */
    @JvmStatic
    external fun setupGuideJson(providerId: String, localeTag: String): String

    /** USD 十进制字符串 → 显示币种字符串。币种符号、位数、locale 都在 Swift 侧。 */
    @JvmStatic
    external fun formatUsd(raw: String, currency: String, localeTag: String): String

    /** 月日档日期（zh「8月20日」/ en "August 20"）。 */
    @JvmStatic
    external fun formatMonthAndDay(millis: Long, localeTag: String): String

    @JvmStatic
    external fun computeDashboardJson(
        snapshotsJson: String,
        subscriptionsJson: String,
        nowMillis: Long,
        currency: String,
        localeTag: String,
        filterJson: String,
    ): String

    /** 详情页历史图视图状态（点/窗口/桶序列/推断段/轴刻度值）。Kotlin 只渲染。 */
    @JvmStatic
    external fun historyChartJson(
        providerId: String,
        snapshotsJson: String,
        range: String,
        nowMillis: Long,
        offset: Int,
        spanLookback: Boolean,
    ): String

    @JvmStatic
    external fun spendBreakdownJson(
        linesJson: String,
        grouping: String,
        currency: String,
        localeTag: String,
    ): String

    @JvmStatic
    external fun postTipJson(payloadJson: String): String

    @JvmStatic
    external fun transferCodeGenerate(): String

    /** 用户敲的转移码规范化。不合法回空串。字母表只在 `MeterCore/TransferCode` 一份。 */
    @JvmStatic
    external fun transferCodeNormalize(input: String): String

    @JvmStatic
    external fun postUsageJson(payloadJson: String): String

    @JvmStatic
    external fun postFeedbackJson(payloadJson: String): String

    @JvmStatic
    external fun inboxCreateJson(): String

    @JvmStatic
    external fun inboxDeleteJson(readKey: String): String

    @JvmStatic
    external fun inboxReadingsJson(readKey: String): String

    @JvmStatic
    external fun inboxListKeysJson(readKey: String): String

    @JvmStatic
    external fun inboxMintKeyJson(readKey: String, label: String): String

    @JvmStatic
    external fun inboxRevokeKeyJson(readKey: String, keyID: String): String

    @JvmStatic
    external fun designSeedJson(nowMillis: Long): String

    @JvmStatic
    external fun fetchProviderJson(
        providerId: String,
        fieldsJson: String,
        nowMillis: Long,
    ): String

    private const val JNI_LIB = "libMeterCoreJNI.so"
}
